README.md
# dart_validator   [![pub package](https://img.shields.io/badge/-0.0.1-brightgreen.svg)](https://pub.dartlang.org/packages/weui)

#### dart数据校验工具, 适用于表单数据校验, Map数据校验

## 安装
```
dart_validator: 0.0.1
```

## 使用

```dart
// 引入
import 'package:dart_validator/dart_validator.dart';

// 定义规则
final Map<String, List<Schema>> descriptor = {
  'name': [
    Schema(
      type: SchemaType.required,
      message: '名称不能为空',
    ),
  ],
  'age': [
    Schema(
      type: SchemaType.required,
      message: '年龄不能为空',
    ),
  ],
  'sex': [
    Schema(
      type: SchemaType.required,
      message: '性别不能为空',
    ),
  ],
  'phone': [
    Schema(
      type: SchemaType.required,
      message: '手机号不能为空',
    ),
    Schema(
      type: SchemaType.phone,
      message: '手机号格式错误',
    ),
  ],
};

// 实例化
final validator = Validator(descriptor);

// 数据
final data = {
  'name': '123',
  'age': '13',
  'sex': '男',
  'phone': '15818440278',
};

// 调用校验
validator.validate(
  data,
  // 失败
  fail: (message, key) {
    print('$key: $message');
  },
  // 成功
  success: () {
    print('校验成功');
  }
);
```

## 自定义校验和异步校验

```dart
import 'package:dart_validator/dart_validator.dart';

// 定义规则
final Map<String, List<Schema>> descriptor = {
  'name': [
    Schema(
      validator: (key, value, callback) {
        // 定时器模拟请求接口
        Future.delayed(Duration(milliseconds: 2000), () {
          callback(value == '张三' ? '输入的名称不是张三' : null);
        });
      }
    ),
  ],
};


// 实例化
final validator = Validator(descriptor);

// 数据
final data = {
  'name': '李四',
};

// 调用校验
validator.validate(
  data,
  // 失败
  fail: (message, key) {
    print('$key: $message');
  },
  // 成功
  success: () {
    print('校验成功');
  }
);
```
#### `Schema`中的`validator`必须调用`callback`函数, 如果数据校验失败传入提示文字, 成功则传入null

CHANGELOG.md
## 0.0.1
- 第一个版本发布，实现验证功能


lib/src/validator.dart
import 'package:dart_validator/src/regexs.dart';
import 'package:dart_validator/src/schema.dart';
export 'package:dart_validator/src/schema.dart';

// 成功
typedef _Success = void Function();
// 失败
typedef _Fail = void Function(String message, String key);

class Validator {
  // 规则
  final Map<String, List<Schema>> rules;
  // 验证
  Function() _validate;
  // 数据
  Map<String, dynamic> _data = {};
  // 成功回调
  _Success _success;
  // 失败回调
  _Fail _fail;

  Validator(this.rules) {
    final methods = _beatFlat(rules);
    final list = _stringChain(methods);
    _validate = list[0];
  }

  // 手机
  static bool isPhone(String phone) => RegExp(Regexs.phone).hasMatch(phone);
  // 邮箱
  static bool isEmail(String email) => RegExp(Regexs.email).hasMatch(email);
  // ip
  static bool isIp(String ip) => RegExp(Regexs.ip).hasMatch(ip);
  // url
  static bool isUrl(String url) => RegExp(Regexs.url).hasMatch(url);
  // 数字
  static bool isNumber(String number) => RegExp(Regexs.number).hasMatch(number);

  // 拍平验证函数
  List<Function(Function next)> _beatFlat(Map<String, List<Schema>> rules) {
    final List<Function(Function next)> list = [];

    for (var key in rules.keys) {
      final List<Schema> rules = this.rules[key];

      for (var rule in rules) {
        list.add((next) {
          // value
          final value = _data[key];

          // 判断是否是自定义验证
          if (rule.type != null) {
            bool status = false;
            // 判断类型
            switch (rule.type) {
              case SchemaType.required:
                status = value == null ? false : (value is String && value != '');
                break;
              case SchemaType.phone:
                status = isPhone(value);
                break;
              case SchemaType.email:
                status = isEmail(value);
                break;
              case SchemaType.ip:
                status = isIp(value);
                break;
              case SchemaType.url:
                status = isUrl(value);
                break;
              case SchemaType.number:
                status = isNumber(value);
                break;
            }

            // 判断结果
            if (status) {
              next();
            } else if (_fail != null) {
              _fail(rule.message, key);
            }
          } else {
            // 自定义验证
            rule.validator(key, value, (String message) {
              if (message == null) {
                next();
              } else if (_fail != null) {
                _fail(message, key);
              }
            });
          }
        });
      }
    }

    return list;
  }

  // 串联
  List<Function()> _stringChain(List<Function(Function next)> params) {
    final List<Function()> list = [
      // 最终成功的回调
      () {
        if (_success != null) _success();
      }
    ];

    for (var next in params.reversed) {
      final last = list.last;
      list.add(() {
        next(last);
      });
    }

    return list.reversed.toList();
  }

  // 验证
  validate(Map<String, dynamic> data, { _Fail fail, _Success success }) {
    _data = data;
    _fail = fail;
    _success = success;
    _validate();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:weui/weui.dart';
import 'package:dart_validator/dart_validator.dart';

void main() => runApp(MyApp());

final Map<String, List<Schema>> descriptor = {
  'name': [
    Schema(
      type: SchemaType.required,
      message: '名称不能为空',
    ),
  ],
  'age': [
    Schema(
      type: SchemaType.required,
      message: '年龄不能为空',
    ),
    Schema(
      validator: (key, value, callback) {
        callback((double.parse(value)) < 18 ? '未成年不能通过' : null);
      }
    )
  ],
  'sex': [
    Schema(
      type: SchemaType.required,
      message: '性别不能为空',
    ),
  ],
  'phone': [
    Schema(
      type: SchemaType.required,
      message: '手机号不能为空',
    ),
    Schema(
      type: SchemaType.phone,
      message: '手机号格式错误',
    ),
  ],
};

final validator = Validator(descriptor);

class MyApp extends StatelessWidget {
  final data = {
    'name': '123',
    'age': '13',
    'sex': '男',
    'phone': '15818440278',
  };

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'validator',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('validator'),
        ),
        body: Builder(
          builder: (BuildContext context) {
            return Center(
              child: MaterialButton(
                color: Colors.blue,
                textColor: Colors.white,
                onPressed: () {
                  validator.validate(
                    data,
                    fail: (message, key) {
                      WeToast.info(context)('$key: $message');
                    },
                    success: () {
                      WeToast.success(context)(
                        message: '校验成功'
                      );
                    }
                  );
                },
                child: Text('验证'),
              ),
            );
          },
        )
      ),
    );
  }
}
